url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/ExecStageARMv5.java

```diff
@@ -0,0 +1,4246 @@
+                        String.format("ldm(1) [%08x]", paddr));
+                return;
+            }
+            setReg(i, read32_a32(paddr));
+            vaddr += 4;
+        }
+        //r15
+        if (BitOp.getBit32(rlist, 15)) {
+            int v;
+
+            paddr = getMMU().translate(vaddr, 4, false, getCPSR().isPrivMode(), true);
+            if (getMMU().isFault()) {
+                getMMU().clearFault();
+                return;
+            }
+
+            if (!tryRead_a32(paddr, 4)) {
+                raiseException(ARMv5.EXCEPT_ABT_DATA,
+                        String.format("ldm(1) [%08x]", paddr));
+                return;
+            }
+            v = read32_a32(paddr);
+
+            setPC(v & 0xfffffffe);
+            getCPSR().setTBit(BitOp.getBit32(v, 0));
+            vaddr += 4;
+        }
+
+        if (w) {
+            setReg(rn, getReg(rn) + len);
+        }
+    }
+
+    /**
+     * ロードマルチプル命令。
+     *
+     * レジスタリストに r15(pc) を入れることはできません。
+     * 現在のモードにかかわらず、ユーザモードのレジスタにロードします。
+     *
+     * @param inst ARM 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executeLdm2(InstructionARM inst, boolean exec) {
+        int rn = inst.getRnField();
+        int rlist = inst.getRegListField();
+        int vaddr, paddr, v, mod;
+
+        if (!exec) {
+            printDisasm(inst,
+                    String.format("ldm%s%s",
+                            inst.getCondFieldName(),
+                            inst.getPUFieldName()),
+                    String.format("%s, {%s}^",
+                            getRegName(rn),
+                            inst.getRegListFieldName()));
+            return;
+        }
+
+        if (!inst.satisfiesCond(getCPSR())) {
+            return;
+        }
+
+        //r15 以外、なお r15 は指定不可のため処理なし
+        vaddr = getAddrMode4StartAddress(inst.getPUField(), rn, rlist);
+        for (int i = 0; i < 15; i++) {
+            if ((rlist & (1 << i)) == 0) {
+                continue;
+            }
+
+            paddr = getMMU().translate(vaddr, 4, false, getCPSR().isPrivMode(), true);
+            if (getMMU().isFault()) {
+                getMMU().clearFault();
+                return;
+            }
+
+            if (!tryRead_a32(paddr, 4)) {
+                raiseException(ARMv5.EXCEPT_ABT_DATA,
+                        String.format("ldm(2) [%08x]", paddr));
+                return;
+            }
+            //必ずユーザモードのレジスタをロードする
+            v = read32_a32(paddr);
+            vaddr += 4;
+
+            mod = getCPSR().getMode();
+            getCPSR().setMode(PSR.MODE_USR);
+            setReg(i, v);
+            getCPSR().setMode(mod);
+        }
+    }
+
+    /**
+     * ロードマルチプル命令。
+     *
+     * レジスタリストに必ず r15(pc) を入れなければなりません。
+     * CPSR にカレントモードの SPSR をコピーします。
+     *
+     * @param inst ARM 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executeLdm3(InstructionARM inst, boolean exec) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 3526-3547 into a separate method.